### DISCUSSION TITLE
Panel layout and actual layout

### DISCUSSION THREAD

**Discussion by Bumblebee001** [GrandOrgue/GoOdf#64] [General]
The two do not match.

This is the panel lay out from GOODF. Note the labels on the stop tabs. All nicely arranged.
![image](https://github.com/larspalo/GOODF/assets/6999990/d28ccbbd-d135-4991-8eca-9db5524e28e3)

On loading the organ, I was horrified..... (well almost)
![image](https://github.com/larspalo/GOODF/assets/6999990/083d9528-6fea-4c57-a8e3-66e1750b047b)

**Comment by Bumblebee001** [comment]
I solved the issue for my case by further adjustments to text settings (width and position from Left) saving, reloading and correcting further as necessary. However, I would expect the layout on GOODF to be accurate enough to show what the loaded organ would look like.

It might be a small thing that could have been overlooked. I do find odd settings I don't recall inputting myself as such... yet there they were eg a keyboard with 59 keys (instead of 61).... How on earth that came about is anybody's best guess! I know by now that a full keyboard has 61 keys and I found the error accidentally whilst working on adding voicing data to the odf manually.

**Comment by larspalo** [comment]
Ok, in the images above you can easily see that the GrandOrgue loading has a scaled window. Due to rounding issues, the scaling that GrandOrgue does of the panels there can be somewhat erratic (not pixel perfect) and some aliasing can happen to the fonts too. Only at a GrandOrgue scaling of 100% will the layout exactly match, and this is a conscious design choice I made when creating the GOODF panel display.

> However, I would expect the layout on GOODF to be accurate enough to show what the loaded organ would look like.

It does reasonably well, when GrandOrgue uses no scaling of the panel but display it as 100% of original (actual) size.

I'll eventually submit an issue report (feature request to GrandOrgue, or even a PR) for a displayed measure of the scaling that GO does and a way to force a 100% scaling - and thus a pixel perfect layout.

The keyboard numbers I've got absolutely no explanation for, I've never encountered that myself (yet). Until you can give me an example that I can re-create the issue with, I'll consider the human error to be the most likely cause...

**Comment by Bumblebee001** [comment]
I was always under the impression that GO was not able to scale. I had
pointed this issue out in the past and it was always ruled out. So that’s
one new to me. As for the odd number of 59 keys I couldn’t tell you. It was
an accidental discovery. As soon as I caught a glimpse of keyboard 59 I
said to myself where did that come from! (I censored the expletives 🤣🤣🤣).

**Reply by larspalo** [reply]
Yes, GrandOrgue does some automatic scaling if the panel is too large for the current display, and also it's possible to drag the corners or sides of the window to adjust size. The scaling of GO really isn't so good as the rounding errors will produce visible faults in the display.

Shit happens... Just keep an eye open if it happens again that the keyboard range changes and if there's a method to reproduce a possible bug.

**Comment by davidgritter** [comment]
I have had a similar thing happen.  I was editing a panel that had been defined in hauptwerk and converted to Grandorgue by ODFedit.  I noticed in the  GOODF panel display that the manual keys did not fill the space between keycheeks completely, unless I adjusted the key spacings to be the same as the dimensions of the bitmap width for each key. In that case it fit exactly and looked correct  However, when loading this in GrandOrgue the keyboard now extended well past the right keycheek and there were spaced between some keys.   The original spacings in the ODF result in the correct filling of the keyboard in GrandOrgue regardless of manual changing the size of the panel window.
